src/notebookConverter.ts

```diff
@@ -511,203 +590,264 @@ export class NotebookConverter implements Disposable {
+      const startLoc = concatDocument.locationAt(range.start);
+      const endLoc = concatDocument.locationAt(range.end);
+      return {
+        uri: startLoc.uri,
+        range: new Range(startLoc.range.start, endLoc.range.end),
+      };
+    }
+    return {
+      uri,
+      range,
+    };
+  }
+
+  private toOutgoingContentChangeEvent(
+    cell: TextDocument,
+    ev: TextDocumentContentChangeEvent
+  ) {
+    return {
+      range: this.toOutgoingRange(cell, ev.range),
+      rangeLength: ev.rangeLength,
+      rangeOffset: this.toOutgoingOffset(cell, ev.rangeOffset),
+      text: ev.text,
+    };
+  }
+
+  private onDidOpenNotebook(doc: NotebookDocument) {
+    if (this.notebookFilter.test(doc.uri.fsPath)) {
+      this.getTextDocumentWrapper(doc.uri);
+      const key = NotebookConverter.getDocumentKey(doc.uri);
+      this.pendingCloseDocuments.delete(key);
+    }
+  }
+
+  private onDidCloseNotebook(doc: NotebookDocument) {
+    if (this.notebookFilter.test(doc.uri.fsPath)) {
+      const key = NotebookConverter.getDocumentKey(doc.uri);
+      const wrapper = this.getTextDocumentWrapper(doc.uri);
+      this.activeDocuments.delete(key);
+      this.activeDocumentsOutgoingMap.delete(
+        NotebookConverter.getDocumentKey(wrapper.uri)
+      );
+      this.pendingCloseDocuments.set(key, wrapper);
+      wrapper.dispose();
+    }
+  }
+
+  private getWrapperFromOutgoingUri(
+    outgoingUri: Uri
+  ): NotebookConcatDocument | undefined {
+    return this.activeDocumentsOutgoingMap.get(
+      NotebookConverter.getDocumentKey(outgoingUri)
+    );
+  }
+
+  private getTextDocumentWrapper(
+    cell: TextDocument | Uri
+  ): NotebookConcatDocument {
+    const uri = cell instanceof Uri ? <Uri>cell : cell.uri;
+    const key = NotebookConverter.getDocumentKey(uri);
+    let result = this.activeDocuments.get(key);
+    if (!result) {
+      const doc = this.api.notebookDocuments.find((n) =>
+        this.fs.arePathsSame(uri.fsPath, n.uri.fsPath)
+      );
+      if (!doc) {
+        throw new Error(`Invalid uri, not a notebook: ${uri.fsPath}`);
+      }
+      result = new NotebookConcatDocument(doc, this.api, this.cellSelector);
+      this.disposables.push(result);
+      result.onCellsChanged(
+        (e) => this.onDidChangeCellsEmitter.fire(e),
+        undefined,
+        this.disposables
+      );
+      this.activeDocuments.set(key, result);
+      this.activeDocumentsOutgoingMap.set(
+        NotebookConverter.getDocumentKey(result.uri),
+        result
+      );
+    }
+    return result;
+  }
+
+  private getConcatDocument(
+    cell: TextDocument | Uri
+  ): IConcatTextDocument | undefined {
+    return this.getTextDocumentWrapper(cell)?.concatDocument;
+  }
+
+  private filterMagics(
+    wrapper: NotebookConcatDocument,
+    value: Diagnostic
+  ): boolean {
+    // Get the code at the range
+    const text = wrapper.getText(value.range);
+
+    // Only skip diagnostics on the front of the line (spacing?)
+    if (
```

Comment:
e.g. `/^\s*\%/` ?

Reply:
No I don't think we want to remove ones that aren't at the front of the line.